<h1>BI Question solutons are calculated in this notebook</h1>

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt
from io import StringIO
import re

In [2]:
#Reading the Merged Dataset that was created after data cleaning
df = pd.read_csv("mergedData.csv",encoding='latin1')

In [3]:
df.head()

Client ID  Age  Gender    County  Income Group  \
0          1   36       1      cork   10001-40000   
1          2   43       1     cavan       0-10000   
2          3   32       0    dublin   10001-40000   
3          4   52       1     louth   40001-60000   
4          5   63       0  kilkenny  60001-100000   

   Average_CA_transaction_amount  No_of_products_held  Held Loan previously  \
0                             58                    4                     1   
1                           2663                    4                     0   
2                             46                    2                     0   
3                              0                    2                     1   
4                            126                    1                     0   

   Num Transactions  Last TXN Amount  Merchant Code  \
0               0.0              NaN            NaN   
1              17.0            83.66         7211.0   
2              25.0           526.18         3667.0   
3              13.0            70.68         5712.0   
4              39.0           259.07         5999.0   

                 Last Transaction Narrative  Loan Flag  
0                                       NaN          0  
1     THE BRIDGE LAUNDRY       WICKLOW TOWN          0  
2  LUXOR HOTEL/CASINO       LAS VEGAS    NV          0  
3     HARVEY NORMAN            CARRICKMINES          0  
4      PAYPAL *PETEWOODWAR      35314369001          0

In [4]:
#Taking only columns required to answer the Business Intellegence questions
bi_df = df[["Client ID","Age" ,"Gender", "Income Group", "Average_CA_transaction_amount", "No_of_products_held","Held Loan previously","Num Transactions","Loan Flag"]]

In [5]:
bi_df.head()

Client ID  Age  Gender  Income Group  Average_CA_transaction_amount  \
0          1   36       1   10001-40000                             58   
1          2   43       1       0-10000                           2663   
2          3   32       0   10001-40000                             46   
3          4   52       1   40001-60000                              0   
4          5   63       0  60001-100000                            126   

   No_of_products_held  Held Loan previously  Num Transactions  Loan Flag  
0                    4                     1               0.0          0  
1                    4                     0              17.0          0  
2                    2                     0              25.0          0  
3                    2                     1              13.0          0  
4                    1                     0              39.0          0

In [6]:
#Checking the basic statistics of the columns
bi_df.describe()

Client ID           Age        Gender  Average_CA_transaction_amount  \
count  10000.00000  10000.000000  10000.000000                   10000.000000   
mean    5000.50000     44.347000      0.509000                     444.829200   
std     2886.89568     14.849857      0.499944                     877.434779   
min        1.00000     20.000000      0.000000                       0.000000   
25%     2500.75000     32.000000      0.000000                       0.000000   
50%     5000.50000     44.000000      1.000000                      30.000000   
75%     7500.25000     57.000000      1.000000                     402.000000   
max    10000.00000    100.000000      1.000000                    9470.000000   

       No_of_products_held  Held Loan previously  Num Transactions  \
count         10000.000000          10000.000000      10000.000000   
mean              3.002700              0.249800         18.292900   
std               1.410068              0.432919         22.962819   
min               0.000000              0.000000          0.000000   
25%               2.000000              0.000000          0.000000   
50%               3.000000              0.000000         10.000000   
75%               4.000000              0.000000         30.000000   
max               5.000000              1.000000        100.000000   

          Loan Flag  
count  10000.000000  
mean       0.021100  
std        0.143725  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [7]:
bi_df.dtypes

Client ID                          int64
Age                                int64
Gender                             int64
Income Group                      object
Average_CA_transaction_amount      int64
No_of_products_held                int64
Held Loan previously               int64
Num Transactions                 float64
Loan Flag                          int64
dtype: object

In [8]:
#Checking for null values
bi_df.isnull().sum()

Client ID                        0
Age                              0
Gender                           0
Income Group                     0
Average_CA_transaction_amount    0
No_of_products_held              0
Held Loan previously             0
Num Transactions                 0
Loan Flag                        0
dtype: int64

In [9]:
#Checking for number of clients in the dataset above age 50 are as follows:
bi_df[bi_df.Age>50]["Client ID"].count()

3816

<h2>Business Intelligence Questions and Answers</h2>

<h3>Question 1: How Many Customers above 50 years old have taken up a loan?</h3>

In [10]:
#Filtering age and Loan flag together with count
(bi_df[bi_df.Age>50]["Loan Flag"] == 1).astype(int).sum(axis = 0)

98

In [11]:
(bi_df[bi_df.Age>50]["Loan Flag"] == 0).astype(int).sum(axis = 0)

3718

<h3> Solution: 98 clinets have taken up loan that were above 50 years of Age</h3>

<h3>Question 2: How Many Females aged 30 to 40 have more than 2 products?</h3>

In [12]:
#Filtered dataset for females between age 30 - 40 that have more than 2 products
bi_df[(bi_df["Gender"] == 0) & (bi_df['Age'].between(30, 40, inclusive=False)) & (bi_df['No_of_products_held'] >2)].head()

Client ID  Age  Gender  Income Group  Average_CA_transaction_amount  \
14         15   37       0   10001-40000                             72   
33         34   32       0  60001-100000                           9470   
62         63   31       0       0-10000                              0   
71         72   32       0   40001-60000                              0   
77         78   36       0  60001-100000                              0   

    No_of_products_held  Held Loan previously  Num Transactions  Loan Flag  
14                    5                     0               0.0          0  
33                    4                     0               0.0          0  
62                    4                     0              40.0          0  
71                    4                     1               0.0          0  
77                    5                     1              64.0          0

In [13]:
#Count of females between age 30 - 40 that have more than 2 products
bi_df[(bi_df["Gender"] == 0) & (bi_df['Age'].between(30, 40, inclusive=False)) & (bi_df['No_of_products_held'] >2)]["No_of_products_held"].count()

513

<h3> Solution: 513 female clients aged between 30 - 40 years old have more than 2 (non inclusive) that were above 50</h3>

<h3>Question 3: What is the average number of Current Account(CA) Transactions for males who had a previous Loans?</h3>

In [14]:
#Avg num of CA Transactions for males who had previous loans
bi_df[(bi_df["Gender"] == 1) & (bi_df["Held Loan previously"] == 1)]["Num Transactions"].mean()

18.646869983948637

<h3> Solution: 19 is the average number of CA transactions for males who has previous loans</h3>

<h3>Question 4: How many females did not have a previous loans and who are aged </h3>
   <h4> Less than 20</h4>
    <h4>21 to 40 </h4>
    <h4>40+ </h4>


In [15]:
#females with age les than 20 (inclusive of 20)
bi_df[(bi_df["Gender"] == 0) & (bi_df["Held Loan previously"] == 0) & (bi_df["Age"] <=20)]["Client ID"].count()

119

In [16]:
#females with age between 21 and 40 ( both inclusive)
bi_df[(bi_df["Gender"] == 0) & (bi_df["Held Loan previously"] == 0) & (bi_df['Age'].between(21, 40, inclusive=True))]["Client ID"].count()

1535

In [17]:
#females with age above 40 years old (excluding 40)
bi_df[(bi_df["Gender"] == 0) & (bi_df["Held Loan previously"] == 0) & (bi_df["Age"] >40)]["Client ID"].count()

2004

<h3> Assumptions:</h3>
    <h4>Converted negetive values in the number of products held column to postive</h4>
    <h4>Set 100 years as limit to the Age column</h4>